typing/printtyp.ml

```diff
@@ -263,99 +238,61 @@ let with_arg id f =
 let fuzzy_id namespace id = namespace = Module && S.mem (Ident.name id) !fuzzy
 
 let with_hidden ids f =
-  let update m id = S.add (Ident.name id.ident) m in
-  protect_refs [ R(protected, List.fold_left update !protected ids)] f
-
-let pervasives_name namespace name =
-  if not !enabled then Out_name.create name else
-  match M.find name (get namespace) with
-  | Associated_to_pervasives r -> r
-  | Need_unique_name _ -> Out_name.create (pervasives name)
-  | Uniquely_associated_to (id',r) ->
-      let hid, map = add_hid_id id' Ident.Map.empty in
-      Out_name.set r (human_unique hid id');
-      Conflicts.collect_explanation namespace hid id';
-      set namespace @@ M.add name (Need_unique_name map) (get namespace);
-      Out_name.create (pervasives name)
-  | exception Not_found ->
-      let r = Out_name.create name in
-      set namespace @@ M.add name (Associated_to_pervasives r) (get namespace);
-      r
-
-(** Lookup for preexisting named item within the current {!printing_env} *)
-let env_ident namespace name =
-  if S.mem name !protected then None else
-  match Namespace.lookup namespace name with
-  | Pident id -> Some id
-  | _ -> None
-  | exception Not_found -> None
+  let update m id = M.add (Ident.name id.ident) id.ident m in
+  let updated = List.fold_left update !bound_in_recursion ids in
+  protect_refs [ R(bound_in_recursion, updated )] f
+
+let indexed_name namespace id =
+  let find namespace id env = match namespace with
+    | Type -> Env.find_type_index id env
+    | Module -> Env.find_module_index id env
+    | Module_type -> Env.find_modtype_index id env
+    | Class -> Env.find_class_index id env
+    | Class_type-> Env.find_cltype_index id env
+    | Other -> None
+  in
+  let rec_bound = M.find_opt (Ident.name id) !bound_in_recursion in
+  match in_printing_env (find namespace id), rec_bound with
+  | None, None
+    (* This case is potentially problematic, it might indicate that
+       the identifier id is not defined in the environment, while there
+       are other identifiers in scope that share the same name.
+       Currently, this kind of partially incoherent environment happens
+       within functor error messages where the left and right hand side
+       have a different views of the environment at the source level.
+       Printing the source-level name seems like a reasonable compromise
+       in this situation however.*)
+  | None, Some _ | Some 0, None -> Ident.name id
+  | Some n, Some p ->
+      if Ident.same p id then
+        Ident.name id
+      else
+        String.concat "/" [Ident.name id; string_of_int (2 + n)]
```

Comment:
I would find the handling of `rec_bound` easier to understand if it was formulated as an information used to transform/fix the indexed name, rather than a parallel information used for printing in addition to the indexed name. (This comes from the view that rec_bound business should not be necessary if the environment correctly had those extra bindings in it, so we are fixing the environment rather than changing the printing.) This would in fact be similar to how `env_ident` was designed in the previous version.

Sketch:

```ocaml
let index = in_printing_env (find namespace id) in
let index =
  (* The environment we are passed does not countain recursive occurrences,
     which should have precedence if they exist. Fix the index if necessary. *)
  match M.find_opt (Ident.name id) !bound_in_recursion with
  | None -> index
  | Some rec_ident ->
    if Ident.same rec_ident id
    then
      (* we are in fact a local recursive occurrence *)
      Some 0
    else
      (* we shadow an additional local recursive occurrence *)
      Some (1 + n)
in
match index with
| None -> (* this case ... *) Ident.name id
| Some n -> human_name id n
```